{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p style=\"text-align: center;\" ><font size=\"7\" >Creation of the PePrMInt Dataset</font></p>\n",
    "<hr />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Author**: Thibault Tubiana  \n",
    "**Version**: 1.0  \n",
    "**Last change**: Major refactoring"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **PREPARATION**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "RECALCULATION = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Depandancies\n",
    " --> See `00-SETYP.ipynb` for depandancies and configuration\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparation\n",
    " --> Please run `01-download_files.ipynb` before running this notebook, it countains all the code to download the required files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import importlib\n",
    "\n",
    "#Pandas configuration\n",
    "pd.options.mode.chained_assignment = (\n",
    "    None  # default='warn', remove pandas warning when adding a new column\n",
    ")\n",
    "\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "%config InlineBackend.figure_format ='svg' #better quality figure figure\n",
    "\n",
    "from tqdm.notebook import tnrange, tqdm\n",
    "tqdm.pandas()  # activate tqdm progressbar for pandas apply\n",
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import all global variables and setings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run \"./00-SETUP.ipynb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Choose here** if a **full recomputation** is needed or if loading a checkpoint after the initial structural dataset computation is enough.  \n",
    "  This is usefull if you want "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f58dcf09cc21471aa370a2627ca7581e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "ToggleButton(value=True, description='Recalculation ?', icon='cogs', tooltip='Click for recalculation')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "recalculation_widget = widgets.ToggleButton(\n",
    "    value=RECALCULATION,\n",
    "    description='Recalculation ?',\n",
    "    disabled=False,\n",
    "    button_style='', # 'success', 'info', 'warning', 'danger' or ''\n",
    "    tooltip='Click for recalculation',\n",
    "    icon='cogs' # (FontAwesome names without the `fa-` prefix)\n",
    ")\n",
    "display(recalculation_widget)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Home/ii/alexanderp/tubiana_etal_2022/notebooks'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Instanciating the builder object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'pepr2ds.builder.Builder' from '/Home/ii/alexanderp/tubiana_etal_2022/pepr2ds/builder/Builder.py'>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n",
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n",
      "INFO: Pandarallel will run on 8 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    }
   ],
   "source": [
    "# import builder.Builder as builderEngine\n",
    "# importlib.reload(builderEngine)\n",
    "\n",
    "# builder = builderEngine.Builder(SETUP, recalculate = recalculation_widget.value, update=False, notebook = True, core=1)\n",
    "\n",
    "import pepr2ds.builder.Builder as builderEngine\n",
    "importlib.reload(builderEngine)\n",
    "builder = builderEngine.Builder(SETUP, recalculate = recalculation_widget.value, update=False, notebook = True, core=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **STRUCTURAL DATASET**\n",
    "## PDB Parsing and computations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first step is to create a dataset based on all PDBs. Here's a quick description\n",
    "1. For each CATH domains, all the PDBS (previously downloaded with `01-download_cathpdb.ipynb`) are **loaded one by one** and **cleaned**.\n",
    "2. The PDBs will be **parsed with biopandas** to be transformed in a **DataFrame**. It will looks like this:  \n",
    "    ![Example biopandas (need internet connectivity)](ressources/biopandas.png) \n",
    "3. The **secondary structure** will be calculated with [**DSSP**  ](https://swift.cmbi.umcn.nl/gv/dssp/)  \n",
    "    ![secondary_structures](ressources/secondaryStructures.png)\n",
    "4. The **Accessible Surface Area** is then computed with [**FREESASA**](http://freesasa.github.io/python/functions.html) (and DSSP..)  \n",
    "    <div><img src=\"ressources/surface-diagram.png\" width=\"400\"></div>\n",
    "5. **Proteins block** are calculated with [**PBxplore**](https://github.com/pierrepo/PBxplore) (check https://github.com/pierrepo/PBxplore)  \n",
    "    <div><img src=\"ressources/PBs.jpg\" width=\"400\"></div>\n",
    "6. Then everything will be merged and saved in the `WORKDIR/dataset`  \n",
    "    ![secondary_structures](ressources/floppy.png)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "29c4ba13989e4c299cbadf97a1a5b22b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "cleaning:   0%|          | 0/10008 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "builder.structure.clean_all_pdbs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".. Processing alfafold structures ..\n",
      "parralel processing 0       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "1       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "2       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "3       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "                              ...                        \n",
      "4087    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4088    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4089    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4090    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4091    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "Length: 4092, dtype: object\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "489586329845409e921486a856410e03",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=512), Label(value='0 / 512'))), HB…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FreeSASA problem /Home/ii/alexanderp/tubiana_etal_2022/notebooks/peprmint/databases/alfafold/PH/extracted/Q6ZR37.pdb Error reading '/Home/ii/alexanderp/tubiana_etal_2022/notebooks/peprmint/databases/alfafold/PH/extracted/Q6ZR37.pdb'.\n",
      "/Home/ii/alexanderp/tubiana_etal_2022/notebooks/peprmint/databases/alfafold/PH/extracted/Q6ZR37.pdb DSSP failed to produce an output\n",
      "PH - Q6ZR37 - local variable 'dssp' referenced before assignment\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FreeSASA:lib/src/structure.c:679: error: input had no valid ATOM or HETATM lines\n",
      "FreeSASA:lib/src/structure.c:686: error: \n",
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".. Processing cath structures ..\n",
      "parralel processing 0       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "1       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "2       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "3       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "4       /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "                              ...                        \n",
      "5911    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "5912    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "5913    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "5914    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "5915    /Home/ii/alexanderp/tubiana_etal_2022/notebook...\n",
      "Length: 5916, dtype: object\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2169d685954d4f0e9bae8005ff8fe560",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=740), Label(value='0 / 740'))), HB…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C2 - 1rsyA00 - '9'\n",
      "PH - 3pegA02 - '1706'\n",
      "SEC14 - 3pegA01 - '1703'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The notebook server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--NotebookApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "NotebookApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 errors\n",
      "fetching experimental method from PDBe\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "96427fd157304fc2a1f270ea55d950e6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6841 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding it into dataset\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a82aad185005419c8ced145254e4440a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=1173483), Label(value='0 / 1173483…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DATASET = builder.structure.build_structural_dataset()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>atom_number</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>x_coord</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>z_coord</th>\n",
       "      <th>occupancy</th>\n",
       "      <th>b_factor</th>\n",
       "      <th>sec_struc</th>\n",
       "      <th>sec_struc_full</th>\n",
       "      <th>prot_block</th>\n",
       "      <th>sasa_rel_dssp</th>\n",
       "      <th>ASA_res_freesasa_florian</th>\n",
       "      <th>RSA_freesasa_florian</th>\n",
       "      <th>ASA_total_freesasa</th>\n",
       "      <th>ASA_mainchain_freesasa</th>\n",
       "      <th>ASA_sidechain_freesasa</th>\n",
       "      <th>RSA_sidechain_freesasa</th>\n",
       "      <th>RSA_total_freesasa_tien</th>\n",
       "      <th>RSA_sidechain_freesasa_tien</th>\n",
       "      <th>sec_struc_segment</th>\n",
       "      <th>pdb</th>\n",
       "      <th>domain</th>\n",
       "      <th>cathpdb</th>\n",
       "      <th>chain</th>\n",
       "      <th>uniprot_acc</th>\n",
       "      <th>data_type</th>\n",
       "      <th>Experimental Method</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [atom_number, atom_name, residue_name, chain_id, residue_number, x_coord, y_coord, z_coord, occupancy, b_factor, sec_struc, sec_struc_full, prot_block, sasa_rel_dssp, ASA_res_freesasa_florian, RSA_freesasa_florian, ASA_total_freesasa, ASA_mainchain_freesasa, ASA_sidechain_freesasa, RSA_sidechain_freesasa, RSA_total_freesasa_tien, RSA_sidechain_freesasa_tien, sec_struc_segment, pdb, domain, cathpdb, chain, uniprot_acc, data_type, Experimental Method]\n",
       "Index: []"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET[DATASET[\"cathpdb\"]== \"1rsyA00\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pdbpath' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [11], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mbiopandas\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpdb\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m PandasPdb\n\u001b[0;32m----> 2\u001b[0m pdbdf \u001b[38;5;241m=\u001b[39m PandasPdb()\u001b[38;5;241m.\u001b[39mread_pdb(pdbpath)\u001b[38;5;241m.\u001b[39mdf[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mATOM\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m#remove duplicated residues.\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m#pdbdf = pdbdf.drop_duplicates(subset=[\"atom_name\",\"chain_id\",\"residue_number\"])\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m#if len(pdbdf.query('residue_number == @lastresid')) == 1:\u001b[39;00m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m#    pdbdf = pdbdf.query(\"residue_number != @lastresid\")\u001b[39;00m\n\u001b[1;32m     14\u001b[0m pdbdf\u001b[38;5;241m.\u001b[39mresidue_name\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pdbpath' is not defined"
     ]
    }
   ],
   "source": [
    "from biopandas.pdb import PandasPdb\n",
    "pdbdf = PandasPdb().read_pdb(pdbpath).df[\"ATOM\"]\n",
    "\n",
    "#remove duplicated residues.\n",
    "#pdbdf = pdbdf.drop_duplicates(subset=[\"atom_name\",\"chain_id\",\"residue_number\"])\n",
    "\n",
    "        #Remove the last residue when sometimes you just have the \"N\"\n",
    "#lastresid = pdbdf.residue_number.values[-1]\n",
    "#if len(pdbdf.query('residue_number == @lastresid')) == 1:\n",
    "#    pdbdf = pdbdf.query(\"residue_number != @lastresid\")\n",
    "\n",
    "    \n",
    "\n",
    "pdbdf.residue_name\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Home/ii/alexanderp/tubiana_etal_2022/notebooks/databases/cath/domains/PLD/raw/3rlgA00.pdb name 'model' is not defined\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'dssp' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [12], line 21\u001b[0m\n\u001b[1;32m     18\u001b[0m structure \u001b[38;5;241m=\u001b[39m p\u001b[38;5;241m.\u001b[39mget_structure(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstruc\u001b[39m\u001b[38;5;124m\"\u001b[39m, pdbpath)\n\u001b[1;32m     19\u001b[0m model \u001b[38;5;241m=\u001b[39m structure[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     20\u001b[0m ssResSmooth, sasa_res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mzip\u001b[39m(\u001b[38;5;241m*\u001b[39m[(builder\u001b[38;5;241m.\u001b[39mstructure\u001b[38;5;241m.\u001b[39msimplify_SS(dssp[x][\u001b[38;5;241m2\u001b[39m]),\n\u001b[0;32m---> 21\u001b[0m                                        dssp[x][\u001b[38;5;241m3\u001b[39m] \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m100\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m dssp\u001b[38;5;241m.\u001b[39mkeys()])\n\u001b[1;32m     23\u001b[0m structure\u001b[38;5;241m.\u001b[39mget_atoms()\n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m#sasa_res\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'dssp' is not defined"
     ]
    }
   ],
   "source": [
    "from Bio.PDB.DSSP import DSSP\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from Bio.SeqUtils import seq3\n",
    "\n",
    "pdbpath = (\"/Home/ii/alexanderp/tubiana_etal_2022/notebooks/databases/cath/domains/PLD/raw/3rlgA00.pdb\")\n",
    "\n",
    "translation_dict = d = {'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "     'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "     'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "     'ALA': 'A', 'VAL':'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'}\n",
    "\n",
    "try:\n",
    "    dssp = DSSP(model, pdbpath, dssp=\"mkdssp\")\n",
    "except Exception as e:\n",
    "    print(pdbpath, e)\n",
    "\n",
    "p = PDBParser(QUIET=False)\n",
    "structure = p.get_structure(\"struc\", pdbpath)\n",
    "model = structure[0]\n",
    "ssResSmooth, sasa_res = zip(*[(builder.structure.simplify_SS(dssp[x][2]),\n",
    "                                       dssp[x][3] * 100) for x in dssp.keys()])\n",
    "\n",
    "structure.get_atoms()\n",
    "\n",
    "#sasa_res\n",
    "len(ssResSmooth)\n",
    "len(list(enumerate(pdbdf.residue_number.unique())))\n",
    "len(pdbdf.atom_number.values)\n",
    "seq = [dssp[key][1] for key in dssp.keys()]\n",
    "\n",
    "new_seq = [a2 for a1,a2 in zip(list(pdbdf.residue_name[1:]), list(pdbdf.residue_name[:-1])) if a1 != a2]\n",
    "if new_seq[-1] != list(pdbdf.residue_name)[-1]:\n",
    "    new_seq.append(list(pdbdf.residue_name)[-1])\n",
    " \n",
    "\n",
    "filterd_seq = []\n",
    "cnt = 0\n",
    "for index, row in pdbdf.iterrows():\n",
    "    if row[\"residue_number\"] > cnt and row[\"atom_name\"]== \"CA\":\n",
    "        filterd_seq.append(row[\"residue_name\"])\n",
    "        try:\n",
    "            print(translation_dict[row[\"residue_name\"]], seq[cnt], cnt, index)\n",
    "        except:\n",
    "            print(\"index out of bounds\")\n",
    "        cnt = cnt + 1\n",
    "        \n",
    "\n",
    "\n",
    "len(filterd_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>atom_number</th>\n",
       "      <th>atom_name</th>\n",
       "      <th>residue_name</th>\n",
       "      <th>chain_id</th>\n",
       "      <th>residue_number</th>\n",
       "      <th>x_coord</th>\n",
       "      <th>y_coord</th>\n",
       "      <th>z_coord</th>\n",
       "      <th>occupancy</th>\n",
       "      <th>b_factor</th>\n",
       "      <th>sec_struc</th>\n",
       "      <th>sec_struc_full</th>\n",
       "      <th>prot_block</th>\n",
       "      <th>sasa_rel_dssp</th>\n",
       "      <th>ASA_res_freesasa_florian</th>\n",
       "      <th>RSA_freesasa_florian</th>\n",
       "      <th>ASA_total_freesasa</th>\n",
       "      <th>ASA_mainchain_freesasa</th>\n",
       "      <th>ASA_sidechain_freesasa</th>\n",
       "      <th>RSA_sidechain_freesasa</th>\n",
       "      <th>RSA_total_freesasa_tien</th>\n",
       "      <th>RSA_sidechain_freesasa_tien</th>\n",
       "      <th>sec_struc_segment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>GLY</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>-22.284</td>\n",
       "      <td>5.561</td>\n",
       "      <td>38.650</td>\n",
       "      <td>1.0</td>\n",
       "      <td>36.15</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>Z</td>\n",
       "      <td>100.00</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CA</td>\n",
       "      <td>GLY</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>-20.924</td>\n",
       "      <td>4.950</td>\n",
       "      <td>38.552</td>\n",
       "      <td>1.0</td>\n",
       "      <td>32.77</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>Z</td>\n",
       "      <td>100.00</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "      <td>GLY</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>-20.075</td>\n",
       "      <td>5.689</td>\n",
       "      <td>37.498</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.53</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>Z</td>\n",
       "      <td>100.00</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>O</td>\n",
       "      <td>GLY</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>-20.343</td>\n",
       "      <td>6.826</td>\n",
       "      <td>37.155</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.00</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>Z</td>\n",
       "      <td>100.00</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>124.802140</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>120.002058</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>N</td>\n",
       "      <td>ASN</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>-19.029</td>\n",
       "      <td>4.997</td>\n",
       "      <td>37.013</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30.07</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>Z</td>\n",
       "      <td>100.00</td>\n",
       "      <td>127.665256</td>\n",
       "      <td>65.469362</td>\n",
       "      <td>127.665256</td>\n",
       "      <td>18.943015</td>\n",
       "      <td>108.722241</td>\n",
       "      <td>105.065946</td>\n",
       "      <td>65.469362</td>\n",
       "      <td>55.754996</td>\n",
       "      <td>C1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2178</th>\n",
       "      <td>2224</td>\n",
       "      <td>CB</td>\n",
       "      <td>LYS</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>-17.746</td>\n",
       "      <td>-3.131</td>\n",
       "      <td>34.802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>36.74</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>f</td>\n",
       "      <td>24.37</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>51.639843</td>\n",
       "      <td>155.206726</td>\n",
       "      <td>95.230535</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>65.765562</td>\n",
       "      <td>C22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2179</th>\n",
       "      <td>2225</td>\n",
       "      <td>CG</td>\n",
       "      <td>LYS</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>-17.381</td>\n",
       "      <td>-4.425</td>\n",
       "      <td>34.027</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.35</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>f</td>\n",
       "      <td>24.37</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>51.639843</td>\n",
       "      <td>155.206726</td>\n",
       "      <td>95.230535</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>65.765562</td>\n",
       "      <td>C22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2180</th>\n",
       "      <td>2226</td>\n",
       "      <td>CD</td>\n",
       "      <td>LYS</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>-18.304</td>\n",
       "      <td>-5.615</td>\n",
       "      <td>34.319</td>\n",
       "      <td>1.0</td>\n",
       "      <td>46.40</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>f</td>\n",
       "      <td>24.37</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>51.639843</td>\n",
       "      <td>155.206726</td>\n",
       "      <td>95.230535</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>65.765562</td>\n",
       "      <td>C22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>2227</td>\n",
       "      <td>CE</td>\n",
       "      <td>LYS</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>-17.805</td>\n",
       "      <td>-6.874</td>\n",
       "      <td>33.618</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50.42</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>f</td>\n",
       "      <td>24.37</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>51.639843</td>\n",
       "      <td>155.206726</td>\n",
       "      <td>95.230535</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>65.765562</td>\n",
       "      <td>C22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2182</th>\n",
       "      <td>2228</td>\n",
       "      <td>NZ</td>\n",
       "      <td>LYS</td>\n",
       "      <td>A</td>\n",
       "      <td>284</td>\n",
       "      <td>-17.526</td>\n",
       "      <td>-8.079</td>\n",
       "      <td>34.606</td>\n",
       "      <td>1.0</td>\n",
       "      <td>54.17</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>f</td>\n",
       "      <td>24.37</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>206.846569</td>\n",
       "      <td>51.639843</td>\n",
       "      <td>155.206726</td>\n",
       "      <td>95.230535</td>\n",
       "      <td>87.646851</td>\n",
       "      <td>65.765562</td>\n",
       "      <td>C22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2128 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      atom_number atom_name residue_name chain_id  residue_number  x_coord  \\\n",
       "0               1         N          GLY        A               2  -22.284   \n",
       "1               2        CA          GLY        A               2  -20.924   \n",
       "2               3         C          GLY        A               2  -20.075   \n",
       "3               4         O          GLY        A               2  -20.343   \n",
       "4               5         N          ASN        A               3  -19.029   \n",
       "...           ...       ...          ...      ...             ...      ...   \n",
       "2178         2224        CB          LYS        A             284  -17.746   \n",
       "2179         2225        CG          LYS        A             284  -17.381   \n",
       "2180         2226        CD          LYS        A             284  -18.304   \n",
       "2181         2227        CE          LYS        A             284  -17.805   \n",
       "2182         2228        NZ          LYS        A             284  -17.526   \n",
       "\n",
       "      y_coord  z_coord  occupancy  b_factor sec_struc sec_struc_full  \\\n",
       "0       5.561   38.650        1.0     36.15         C              -   \n",
       "1       4.950   38.552        1.0     32.77         C              -   \n",
       "2       5.689   37.498        1.0     30.53         C              -   \n",
       "3       6.826   37.155        1.0     29.00         C              -   \n",
       "4       4.997   37.013        1.0     30.07         C              -   \n",
       "...       ...      ...        ...       ...       ...            ...   \n",
       "2178   -3.131   34.802        1.0     36.74         C              -   \n",
       "2179   -4.425   34.027        1.0     40.35         C              -   \n",
       "2180   -5.615   34.319        1.0     46.40         C              -   \n",
       "2181   -6.874   33.618        1.0     50.42         C              -   \n",
       "2182   -8.079   34.606        1.0     54.17         C              -   \n",
       "\n",
       "     prot_block  sasa_rel_dssp  ASA_res_freesasa_florian  \\\n",
       "0             Z         100.00                124.802140   \n",
       "1             Z         100.00                124.802140   \n",
       "2             Z         100.00                124.802140   \n",
       "3             Z         100.00                124.802140   \n",
       "4             Z         100.00                127.665256   \n",
       "...         ...            ...                       ...   \n",
       "2178          f          24.37                206.846569   \n",
       "2179          f          24.37                206.846569   \n",
       "2180          f          24.37                206.846569   \n",
       "2181          f          24.37                206.846569   \n",
       "2182          f          24.37                206.846569   \n",
       "\n",
       "      RSA_freesasa_florian  ASA_total_freesasa  ASA_mainchain_freesasa  \\\n",
       "0               120.002058          124.802140              124.802140   \n",
       "1               120.002058          124.802140              124.802140   \n",
       "2               120.002058          124.802140              124.802140   \n",
       "3               120.002058          124.802140              124.802140   \n",
       "4                65.469362          127.665256               18.943015   \n",
       "...                    ...                 ...                     ...   \n",
       "2178             87.646851          206.846569               51.639843   \n",
       "2179             87.646851          206.846569               51.639843   \n",
       "2180             87.646851          206.846569               51.639843   \n",
       "2181             87.646851          206.846569               51.639843   \n",
       "2182             87.646851          206.846569               51.639843   \n",
       "\n",
       "      ASA_sidechain_freesasa  RSA_sidechain_freesasa  RSA_total_freesasa_tien  \\\n",
       "0                   0.000000                0.000000               120.002058   \n",
       "1                   0.000000                0.000000               120.002058   \n",
       "2                   0.000000                0.000000               120.002058   \n",
       "3                   0.000000                0.000000               120.002058   \n",
       "4                 108.722241              105.065946                65.469362   \n",
       "...                      ...                     ...                      ...   \n",
       "2178              155.206726               95.230535                87.646851   \n",
       "2179              155.206726               95.230535                87.646851   \n",
       "2180              155.206726               95.230535                87.646851   \n",
       "2181              155.206726               95.230535                87.646851   \n",
       "2182              155.206726               95.230535                87.646851   \n",
       "\n",
       "      RSA_sidechain_freesasa_tien sec_struc_segment  \n",
       "0                        0.000000                C1  \n",
       "1                        0.000000                C1  \n",
       "2                        0.000000                C1  \n",
       "3                        0.000000                C1  \n",
       "4                       55.754996                C1  \n",
       "...                           ...               ...  \n",
       "2178                    65.765562               C22  \n",
       "2179                    65.765562               C22  \n",
       "2180                    65.765562               C22  \n",
       "2181                    65.765562               C22  \n",
       "2182                    65.765562               C22  \n",
       "\n",
       "[2128 rows x 23 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'dssp' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [15], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m ids \u001b[38;5;241m=\u001b[39m [r_id[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;28;01mfor\u001b[39;00m c_id,r_id \u001b[38;5;129;01min\u001b[39;00m dssp\u001b[38;5;241m.\u001b[39mkeys()]\n\u001b[1;32m      2\u001b[0m new_data \u001b[38;5;241m=\u001b[39m pdbdf[pdbdf[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresidue_number\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39misin(ids)]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'dssp' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'new_data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [16], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m filterd_seq \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      2\u001b[0m cnt \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m index, row \u001b[38;5;129;01min\u001b[39;00m new_data\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresidue_number\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m>\u001b[39m cnt \u001b[38;5;129;01mand\u001b[39;00m row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124matom_name\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCA\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m      5\u001b[0m         filterd_seq\u001b[38;5;241m.\u001b[39mappend(row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresidue_name\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'new_data' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Home/ii/alexanderp/tubiana_etal_2022/notebooks'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'DATASET' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m DATASET\u001b[38;5;241m.\u001b[39mdata_type\u001b[38;5;241m.\u001b[39munique()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'DATASET' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing protrusion models\n",
    "Based on atom properties of all PDBs (which are in a dataset datastructure), we will compute **convexhull**, **vertices**, **protrusions**, **co-insertables** according Fuglebakk *et al.*, PONe, 2018: https://doi.org/10.1371/journal.pcbi.1006325  \n",
    "![](ressources/protrusions.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... Conputing protrusions .. \n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cb0f5efcd6a84c79bdb07f2aa4c9b076",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(HBox(children=(IntProgress(value=0, description='0.00%', max=1251), Label(value='0 / 1251'))), …"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DATASET = builder.structure.add_protrusions(DATASET)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding structural cluster information\n",
    "\n",
    "To remove redundancy on our dataset (or filter based on redundancy later on), we need cluster informations. For that, we will use <img src=\"ressources/cathlogo.png\" width=\"40\"> clusters\n",
    "\n",
    "In CATH, we can find info about the the sequence cluster.  \n",
    "Note that there is 4 clusters:\n",
    " - S35 : Structure with 35% of sequence idendity\n",
    " - S60 : Structure with 65% of sequence idendity\n",
    " - S95 : Structure with 95% of sequence idendity\n",
    " - S100 : Structure with 100% of sequence idendity\n",
    "\n",
    "But the cluster number is hierarchical actually.  \n",
    "```\n",
    "S35\n",
    " |_S60\n",
    "    |_S95\n",
    "       |_S100\n",
    "```\n",
    "So we need to make a transformation to keep the cluster info. \n",
    "For example, a sequence that is in the cluster 2 at 35%, cluster 4 at 60%, cluster 3 at 95% and cluster 1 at 100% will be `2.4.3.1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'pepr2ds.builder.Builder' from '/Home/ii/alexanderp/tubiana_etal_2022/pepr2ds/builder/Builder.py'>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pepr2ds.builder.Builder as builderEngine\n",
    "importlib.reload(builderEngine)\n",
    "builder = builderEngine.Builder(SETUP, recalculate = recalculation_widget.value, update=False, notebook = True, core=1)\n",
    "\n",
    "DATASET = builder.structure.add_structural_cluster_info(DATASET)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **WORKING WITH SEQUENCES**  \n",
    "Now that we are done with pdb files, with can work on sequences and other external databases\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding uniprot information (origin and uniprot_id)\n",
    "\n",
    "From the PDBid, we will fetch **origin** and **uniprot_id** (like `ASAP1_HUMAN`) from the **uniprot_acc** (`Q9ULH1`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "{'Server': 'nginx/1.17.7', 'Vary': 'Accept-Encoding, Accept, Accept-Encoding, X-UniProt-Release, X-API-Deployment-Date, User-Agent', 'Cache-Control': 'no-cache', 'Content-Type': 'application/json', 'Access-Control-Allow-Credentials': 'true', 'Content-Encoding': 'gzip', 'Access-Control-Expose-Headers': 'Link, X-Total-Results, X-UniProt-Release, X-UniProt-Release-Date, X-API-Deployment-Date', 'X-API-Deployment-Date': '20-September-2022', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Date': 'Fri, 30 Sep 2022 08:37:37 GMT', 'X-UniProt-Release': '2022_03', 'Link': '<https://rest.uniprot.org/idmapping/uniprotkb/results/503b151bbc700d79ceac99e88290eda4a5d5d682?cursor=t49h2mo5cklw&size=500>; rel=\"next\"', 'X-Total-Results': '4388', 'Transfer-Encoding': 'chunked', 'Access-Control-Allow-Origin': '*', 'Connection': 'keep-alive', 'Access-Control-Allow-Methods': 'GET, PUT, POST, DELETE, PATCH, OPTIONS', 'Access-Control-Allow-Headers': 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization', 'X-UniProt-Release-Date': '03-August-2022'}\n",
      "Fetched: 500 / 4388\n",
      "Fetched: 1000 / 4388\n",
      "Fetched: 1500 / 4388\n",
      "Fetched: 2000 / 4388\n",
      "Fetched: 2500 / 4388\n",
      "Fetched: 3000 / 4388\n",
      "Fetched: 3500 / 4388\n",
      "Fetched: 4000 / 4388\n",
      "Fetched: 4388 / 4388\n",
      "[{'from': 'P26256', 'to': {'entryType': 'UniProtKB reviewed (Swiss-Prot)', 'primaryAccession': 'P26256', 'uniProtkbId': 'ANX12_HYDVU', 'entryAudit': {'firstPublicDate': '1992-05-01', 'lastAnnotationUpdateDate': '2022-08-03', 'lastSequenceUpdateDate': '1992-05-01', 'entryVersion': 111, 'sequenceVersion': 1}, 'annotationScore': 3.0, 'organism': {'scientificName': 'Hydra vulgaris', 'commonName': 'Hydra', 'synonyms': ['Hydra attenuata'], 'taxonId': 6087, 'lineage': ['Eukaryota', 'Metazoa', 'Cnidaria', 'Hydrozoa', 'Hydroidolina', 'Anthoathecata', 'Aplanulata', 'Hydridae', 'Hydra']}, 'proteinExistence': '1: Evidence at protein level', 'proteinDescription': {'recommendedName': {'fullName': {'value': 'Annexin-B12'}}, 'alternativeNames': [{'fullName': {'value': 'Annexin XII'}}, {'fullName': {'value': 'Annexin-12'}}]}, 'genes': [{'geneName': {'value': 'ANXB12'}}], 'comments': [{'texts': [{'value': 'Homohexamer'}], 'commentType': 'SUBUNIT'}, {'texts': [{'value': 'A pair of annexin repeats may form one binding site for calcium and phospholipid'}], 'commentType': 'DOMAIN'}, {'texts': [{'value': 'Phosphorylated in vitro on serine(s) and threonine(s) by PKC'}], 'commentType': 'PTM'}, {'texts': [{'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU01245'}, {'evidenceCode': 'ECO:0000305'}], 'value': 'Belongs to the annexin family'}], 'commentType': 'SIMILARITY'}], 'features': [{'type': 'Chain', 'location': {'start': {'value': 1, 'modifier': 'EXACT'}, 'end': {'value': 316, 'modifier': 'EXACT'}}, 'description': 'Annexin-B12', 'featureId': 'PRO_0000067519'}, {'type': 'Repeat', 'location': {'start': {'value': 13, 'modifier': 'EXACT'}, 'end': {'value': 84, 'modifier': 'EXACT'}}, 'description': 'Annexin 1', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU01245'}]}, {'type': 'Repeat', 'location': {'start': {'value': 85, 'modifier': 'EXACT'}, 'end': {'value': 156, 'modifier': 'EXACT'}}, 'description': 'Annexin 2', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU01245'}]}, {'type': 'Repeat', 'location': {'start': {'value': 168, 'modifier': 'EXACT'}, 'end': {'value': 240, 'modifier': 'EXACT'}}, 'description': 'Annexin 3', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU01245'}]}, {'type': 'Repeat', 'location': {'start': {'value': 244, 'modifier': 'EXACT'}, 'end': {'value': 315, 'modifier': 'EXACT'}}, 'description': 'Annexin 4', 'evidences': [{'evidenceCode': 'ECO:0000255', 'source': 'PROSITE-ProRule', 'id': 'PRU01245'}]}, {'type': 'Modified residue', 'location': {'start': {'value': 6, 'modifier': 'EXACT'}, 'end': {'value': 6, 'modifier': 'EXACT'}}, 'description': 'Phosphothreonine; by PKC', 'evidences': [{'evidenceCode': 'ECO:0000255'}]}, {'type': 'Helix', 'location': {'start': {'value': 15, 'modifier': 'EXACT'}, 'end': {'value': 26, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 27, 'modifier': 'EXACT'}, 'end': {'value': 30, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 33, 'modifier': 'EXACT'}, 'end': {'value': 42, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 45, 'modifier': 'EXACT'}, 'end': {'value': 59, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 63, 'modifier': 'EXACT'}, 'end': {'value': 70, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 73, 'modifier': 'EXACT'}, 'end': {'value': 83, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 86, 'modifier': 'EXACT'}, 'end': {'value': 99, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 100, 'modifier': 'EXACT'}, 'end': {'value': 102, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 105, 'modifier': 'EXACT'}, 'end': {'value': 112, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 117, 'modifier': 'EXACT'}, 'end': {'value': 131, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 135, 'modifier': 'EXACT'}, 'end': {'value': 142, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 145, 'modifier': 'EXACT'}, 'end': {'value': 155, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 167, 'modifier': 'EXACT'}, 'end': {'value': 180, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Turn', 'location': {'start': {'value': 181, 'modifier': 'EXACT'}, 'end': {'value': 183, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 184, 'modifier': 'EXACT'}, 'end': {'value': 186, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 189, 'modifier': 'EXACT'}, 'end': {'value': 198, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 201, 'modifier': 'EXACT'}, 'end': {'value': 214, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 215, 'modifier': 'EXACT'}, 'end': {'value': 217, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 219, 'modifier': 'EXACT'}, 'end': {'value': 226, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 229, 'modifier': 'EXACT'}, 'end': {'value': 257, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Beta strand', 'location': {'start': {'value': 259, 'modifier': 'EXACT'}, 'end': {'value': 261, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1AEI'}]}, {'type': 'Helix', 'location': {'start': {'value': 264, 'modifier': 'EXACT'}, 'end': {'value': 274, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Turn', 'location': {'start': {'value': 275, 'modifier': 'EXACT'}, 'end': {'value': 278, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 279, 'modifier': 'EXACT'}, 'end': {'value': 290, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 294, 'modifier': 'EXACT'}, 'end': {'value': 301, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}, {'type': 'Helix', 'location': {'start': {'value': 304, 'modifier': 'EXACT'}, 'end': {'value': 314, 'modifier': 'EXACT'}}, 'description': '', 'evidences': [{'evidenceCode': 'ECO:0007829', 'source': 'PDB', 'id': '1DM5'}]}], 'keywords': [{'id': 'KW-0002', 'category': 'Technical term', 'name': '3D-structure'}, {'id': 'KW-0041', 'category': 'Domain', 'name': 'Annexin'}, {'id': 'KW-0106', 'category': 'Ligand', 'name': 'Calcium'}, {'id': 'KW-0111', 'category': 'Ligand', 'name': 'Calcium/phospholipid-binding'}, {'id': 'KW-0903', 'category': 'Technical term', 'name': 'Direct protein sequencing'}, {'id': 'KW-0597', 'category': 'PTM', 'name': 'Phosphoprotein'}, {'id': 'KW-1185', 'category': 'Technical term', 'name': 'Reference proteome'}, {'id': 'KW-0677', 'category': 'Domain', 'name': 'Repeat'}], 'references': [{'citation': {'id': '1339458', 'citationType': 'journal article', 'authors': ['Schlaepfer D.D.', 'Fisher D.A.', 'Brandt M.E.', 'Bode H.R.', 'Jones J.M.', 'Haigler H.T.'], 'citationCrossReferences': [{'database': 'PubMed', 'id': '1339458'}, {'database': 'DOI', 'id': '10.1016/s0021-9258(19)50123-5'}], 'title': 'Identification of a novel annexin in Hydra vulgaris. Characterization, cDNA cloning, and protein kinase C phosphorylation of annexin XII.', 'publicationDate': '1992', 'journal': 'J. Biol. Chem.', 'firstPage': '9529', 'lastPage': '9539', 'volume': '267'}, 'referencePositions': ['NUCLEOTIDE SEQUENCE [MRNA]', 'PARTIAL PROTEIN SEQUENCE']}, {'citation': {'id': '7477411', 'citationType': 'journal article', 'authors': ['Luecke H.', 'Chang B.T.', 'Maillard W.S.', 'Schlaepfer D.D.', 'Haigler H.T.'], 'citationCrossReferences': [{'database': 'PubMed', 'id': '7477411'}, {'database': 'DOI', 'id': '10.1038/378512a0'}], 'title': 'Crystal structure of the annexin XII hexamer and implications for bilayer insertion.', 'publicationDate': '1995', 'journal': 'Nature', 'firstPage': '512', 'lastPage': '515', 'volume': '378'}, 'referencePositions': ['X-RAY CRYSTALLOGRAPHY (2.8 ANGSTROMS)']}, {'citation': {'id': '10704197', 'citationType': 'journal article', 'authors': ['Cartailler J.-P.', 'Haigler H.T.', 'Luecke H.'], 'citationCrossReferences': [{'database': 'PubMed', 'id': '10704197'}, {'database': 'DOI', 'id': '10.1021/bi992278d'}], 'title': 'Annexin XII E105K crystal structure: identification of a pH-dependent switch for mutant hexamerization.', 'publicationDate': '2000', 'journal': 'Biochemistry', 'firstPage': '2475', 'lastPage': '2483', 'volume': '39'}, 'referencePositions': ['X-RAY CRYSTALLOGRAPHY (1.93 ANGSTROMS) OF MUTANT LYS-105']}], 'uniProtKBCrossReferences': [{'database': 'EMBL', 'id': 'M83736', 'properties': [{'key': 'ProteinId', 'value': 'AAA29206.1'}, {'key': 'Status', 'value': '-'}, {'key': 'MoleculeType', 'value': 'mRNA'}]}, {'database': 'PIR', 'id': 'A42660', 'properties': [{'key': 'EntryName', 'value': 'LUJF12'}]}, {'database': 'RefSeq', 'id': 'NP_001296699.1', 'properties': [{'key': 'NucleotideSequenceId', 'value': 'NM_001309770.1'}]}, {'database': 'RefSeq', 'id': 'XP_002155286.3', 'properties': [{'key': 'NucleotideSequenceId', 'value': 'XM_002155250.3'}]}, {'database': 'PDB', 'id': '1AEI', 'properties': [{'key': 'Method', 'value': 'X-ray'}, {'key': 'Resolution', 'value': '2.80 A'}, {'key': 'Chains', 'value': 'A/B/C/D/E/F=2-316'}]}, {'database': 'PDB', 'id': '1DM5', 'properties': [{'key': 'Method', 'value': 'X-ray'}, {'key': 'Resolution', 'value': '1.93 A'}, {'key': 'Chains', 'value': 'A/B/C/D/E/F=2-316'}]}, {'database': 'PDBsum', 'id': '1AEI', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'PDBsum', 'id': '1DM5', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'AlphaFoldDB', 'id': 'P26256', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'SMR', 'id': 'P26256', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'TCDB', 'id': '1.A.31.1.9', 'properties': [{'key': 'FamilyName', 'value': 'the annexin (annexin) family'}]}, {'database': 'PRIDE', 'id': 'P26256', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'GeneID', 'id': '100203975', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'KEGG', 'id': 'hmg:100203975', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'OMA', 'id': 'KEEYRAM', 'properties': [{'key': 'Fingerprint', 'value': '-'}]}, {'database': 'OrthoDB', 'id': '856254at2759', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'EvolutionaryTrace', 'id': 'P26256', 'properties': [{'key': 'Description', 'value': '-'}]}, {'database': 'Proteomes', 'id': 'UP000694840', 'properties': [{'key': 'Component', 'value': 'Unplaced'}]}, {'database': 'GO', 'id': 'GO:0005509', 'properties': [{'key': 'GoTerm', 'value': 'F:calcium ion binding'}, {'key': 'GoEvidenceType', 'value': 'IEA:InterPro'}]}, {'database': 'GO', 'id': 'GO:0005544', 'properties': [{'key': 'GoTerm', 'value': 'F:calcium-dependent phospholipid binding'}, {'key': 'GoEvidenceType', 'value': 'IEA:UniProtKB-KW'}]}, {'database': 'Gene3D', 'id': '1.10.220.10', 'properties': [{'key': 'EntryName', 'value': '-'}, {'key': 'MatchStatus', 'value': '4'}]}, {'database': 'InterPro', 'id': 'IPR001464', 'properties': [{'key': 'EntryName', 'value': 'Annexin'}]}, {'database': 'InterPro', 'id': 'IPR018502', 'properties': [{'key': 'EntryName', 'value': 'Annexin_repeat'}]}, {'database': 'InterPro', 'id': 'IPR018252', 'properties': [{'key': 'EntryName', 'value': 'Annexin_repeat_CS'}]}, {'database': 'InterPro', 'id': 'IPR037104', 'properties': [{'key': 'EntryName', 'value': 'Annexin_sf'}]}, {'database': 'Pfam', 'id': 'PF00191', 'properties': [{'key': 'EntryName', 'value': 'Annexin'}, {'key': 'MatchStatus', 'value': '4'}]}, {'database': 'PRINTS', 'id': 'PR00196', 'properties': [{'key': 'EntryName', 'value': 'ANNEXIN'}]}, {'database': 'SMART', 'id': 'SM00335', 'properties': [{'key': 'EntryName', 'value': 'ANX'}, {'key': 'MatchStatus', 'value': '4'}]}, {'database': 'SUPFAM', 'id': 'SSF47874', 'properties': [{'key': 'EntryName', 'value': 'SSF47874'}, {'key': 'MatchStatus', 'value': '1'}]}, {'database': 'PROSITE', 'id': 'PS00223', 'properties': [{'key': 'EntryName', 'value': 'ANNEXIN_1'}, {'key': 'MatchStatus', 'value': '3'}]}, {'database': 'PROSITE', 'id': 'PS51897', 'properties': [{'key': 'EntryName', 'value': 'ANNEXIN_2'}, {'key': 'MatchStatus', 'value': '4'}]}], 'sequence': {'value': 'MVVQGTVKPHASFNSREDAETLRKAMKGIGTDEKSITHILATRSNAQRQQIKTDYTTLFGKHLEDELKSELSGNYEAAALALLRKPDEFLAEQLHAAMKGLGTDENALIDILCTQSNAQIHAIKAAFKLLYKEDLEKEIISETSGNFQRLLVSMLQGGRKEDEPVNAAHAAEDAAAIYQAGEGQIGTDESRFNAVLATRSYPQLHQIFHEYSKISNKTILQAIENEFSGDIKNGLLAIVKSVENRFAYFAERLHHAMKGLGTSDKTLIRILVSRSEIDLANIKETFQAMYGKSLYEFIADDCSGDYKDLLLQITGH', 'length': 316, 'molWeight': 35109, 'crc64': '10599869CBA853EC', 'md5': '2BB1D3F1A9250CA373652F1198DBCE09'}, 'extraAttributes': {'countByCommentType': {'SUBUNIT': 1, 'DOMAIN': 1, 'PTM': 1, 'SIMILARITY': 1}, 'countByFeatureType': {'Chain': 1, 'Repeat': 4, 'Modified residue': 1, 'Helix': 19, 'Beta strand': 5, 'Turn': 2}, 'uniParcId': 'UPI0000125B90'}}}]\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c8dedebafa3b48bbbae3e53475b40833",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/9387860 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DATASET = builder.sequence.add_uniprotId_Origin(DATASET)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Starting to add sequences in DATASET\n",
    "\n",
    "We can start to add sequences in our dataset from **PROSITE**. \n",
    "1. We take the multiple alignment from **prosite** and get all sequences.\n",
    "2. for now we keep only sequences that are already in the DATASET (sequences that have a PDB structure)\n",
    "3. we match the residue number of each residues in the PDB, with the residue position in the alignment\n",
    "4. we add all those new data in the DATASET"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> Matching between the structure and the sequence aligned...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5f1c346952b84ea78a5ec68c2adcc2fe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2aa526c978344db7a7407e51e9125d20",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/451 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5b1b31de62304a9a9a9519c598cd2e86",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/191 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fcf40aed304a41b2b016322ab5b88d27",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/337 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "59b75e19f1e84a1eaf7ccd9289ddf315",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/883 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6b1bc08c6cb640ea91bd3a9821f9304d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1354 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5f1082051249450c90149bfe4c1ec5a9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/190 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3286daebf28641c1a4621f9c2c06275d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2273 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f7692cf16db94ed68d460e2ca7e08258",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/207 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3721cebcdd8b44a886fcc2abc4824cca",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/1520 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3eaf5754decf4c138d1061b235398094",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/540 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c48ac39ee01545a7996de3fdfc10f18f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/198 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0cf7450cc31346f093d571fc81f94e55",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/225 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d43e88c827854b3ca2c55759031559fc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/148 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "13131e396879437bacae291b4f3bc140",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/965 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d976094b0ced45859fee6c6b208e6658",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/522 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "DATASET = builder.sequence.match_residue_number_with_alignment_position(DATASET)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding sequence without structures  \n",
    "\n",
    "Of course all sequences doesn't have a structure, but they can be quite usefull for certain statistics (amino acid composition, conservation..), so we add them in our dataset even if they don't have a structure 🙂.  \n",
    "For every amino acid of every sequences, we will consider the residue as a `CA` atom, to match the current dataframe structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "47be52e3b9904b68b3d346f0868a9a7f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PH\n",
      "C2\n",
      "C1\n",
      "PX\n",
      "FYVE\n",
      "BAR\n",
      "ENTH\n",
      "SH2\n",
      "SEC14\n",
      "START\n",
      "C2DIS\n",
      "GLA\n",
      "PLD\n",
      "PLA\n",
      "ANNEXIN\n",
      "> Adding the Sequence data in the dataset...\n"
     ]
    }
   ],
   "source": [
    "DATASET = builder.sequence.add_sequence_in_dataset(DATASET)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding info from uniprot protein sheet\n",
    "\n",
    "Uniprot page can have a lot of usefull data (taxon, origin crossreferences). First we will download all UNIPROT xml sheet for every `uniprotid` we have in our dataset and then we parse it to get the data we want.  \n",
    "For now the data we keep are \n",
    " - [x] `uniprot_id` (for matching)\n",
    " - [x] `uniprot_acc` (for matching)\n",
    " - [x] `location` in the cell\n",
    " - [x] `taxon` for species \n",
    " - [x] `CR:prositeID` Cross reference with PROSITE ID like `PS50003`\n",
    " - [x] `CR:prositeName` for prosite name (like `PH`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d9e1707bc3514eaaaf2d747be39f86d5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6452 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> 0 new files downloaded and 6452 will be reused.\n",
      "> Parsing uniprot files\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "230a583f1ef44c3dabb0545812f84b3c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/6452 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "builder.sequence.download_uniprot_data(DATASET)\n",
    "DATASET = builder.sequence.add_info_from_uniprot(DATASET)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Cluster info \n",
    "\n",
    "To remove the redundancy we can use the \"Uniref\" cluster info. There are 3 differents cluster level:\n",
    "- `uniref50` at 50% of sequence idendity \n",
    "- `uniref90` at 90% of sequence idendity \n",
    "- `uniref100` at 100% of sequence idendity  \n",
    "\n",
    "Every sequences with more than `50`/`90`/`100`% of sequence identity are cluseters together.  \n",
    "More info can be found here https://www.uniprot.org/help/uniref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trying to fetch UniRef50\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "{'Server': 'nginx/1.17.7', 'Vary': 'Accept-Encoding, Accept, Accept-Encoding, X-UniProt-Release, X-API-Deployment-Date, User-Agent', 'Cache-Control': 'no-cache', 'Content-Type': 'application/json', 'Access-Control-Allow-Credentials': 'true', 'Content-Encoding': 'gzip', 'Access-Control-Expose-Headers': 'Link, X-Total-Results, X-UniProt-Release, X-UniProt-Release-Date, X-API-Deployment-Date', 'X-API-Deployment-Date': '20-September-2022', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Date': 'Fri, 30 Sep 2022 10:56:07 GMT', 'X-UniProt-Release': '2022_03', 'Link': '<https://rest.uniprot.org/idmapping/uniref/results/7609811d6dd94c2e44aa172987006bf1e2515798?cursor=t49h2mo5cklw&size=500>; rel=\"next\"', 'X-Total-Results': '6452', 'Transfer-Encoding': 'chunked', 'Access-Control-Allow-Origin': '*', 'Connection': 'keep-alive', 'Access-Control-Allow-Methods': 'GET, PUT, POST, DELETE, PATCH, OPTIONS', 'Access-Control-Allow-Headers': 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization', 'X-UniProt-Release-Date': '03-August-2022'}\n",
      "Fetched: 500 / 6452\n",
      "Fetched: 1000 / 6452\n",
      "Fetched: 1500 / 6452\n",
      "Fetched: 2000 / 6452\n",
      "Fetched: 2500 / 6452\n",
      "Fetched: 3000 / 6452\n",
      "Fetched: 3500 / 6452\n",
      "Fetched: 4000 / 6452\n",
      "Fetched: 4500 / 6452\n",
      "Fetched: 5000 / 6452\n",
      "Fetched: 5500 / 6452\n",
      "Fetched: 6000 / 6452\n",
      "Fetched: 6452 / 6452\n",
      "[{'from': 'P26256', 'to': {'id': 'UniRef50_P26256', 'name': 'Cluster: Annexin-B12', 'updated': '2022-08-03', 'entryType': 'UniRef50', 'commonTaxon': {'scientificName': 'Eumetazoa', 'taxonId': 6072}, 'memberCount': 9, 'organismCount': 4, 'representativeMember': {'memberIdType': 'UniProtKB ID', 'memberId': 'ANX12_HYDVU', 'organismName': 'Hydra vulgaris (Hydra) (Hydra attenuata)', 'organismTaxId': 6087, 'sequenceLength': 316, 'proteinName': 'Annexin-B12', 'accessions': ['P26256'], 'uniref90Id': 'UniRef90_P26256', 'uniref100Id': 'UniRef100_P26256', 'uniparcId': 'UPI0000125B90', 'sequence': {'value': 'MVVQGTVKPHASFNSREDAETLRKAMKGIGTDEKSITHILATRSNAQRQQIKTDYTTLFGKHLEDELKSELSGNYEAAALALLRKPDEFLAEQLHAAMKGLGTDENALIDILCTQSNAQIHAIKAAFKLLYKEDLEKEIISETSGNFQRLLVSMLQGGRKEDEPVNAAHAAEDAAAIYQAGEGQIGTDESRFNAVLATRSYPQLHQIFHEYSKISNKTILQAIENEFSGDIKNGLLAIVKSVENRFAYFAERLHHAMKGLGTSDKTLIRILVSRSEIDLANIKETFQAMYGKSLYEFIADDCSGDYKDLLLQITGH', 'length': 316, 'molWeight': 35109, 'crc64': '10599869CBA853EC', 'md5': '2BB1D3F1A9250CA373652F1198DBCE09'}}, 'seedId': 'A0A8K0ENN7', 'memberIdTypes': ['UniProtKB Unreviewed (TrEMBL)', 'UniProtKB Reviewed (Swiss-Prot)', 'UniParc'], 'members': ['P26256,0', 'C3ZND3,1', 'A0A8J9ZKK6,1', 'A0A8J9ZKK9,1', 'A0A8J9ZLY0,1', 'A0A8K0ENN7,1', 'UPI001EE5EDC9,3', 'UPI00160432D4,3', 'UPI0015FF0BE7,3'], 'organisms': [{'scientificName': 'Hydra vulgaris', 'commonName': 'Hydra', 'taxonId': 6087}, {'scientificName': 'Branchiostoma floridae', 'commonName': 'Florida lancelet', 'taxonId': 7739}, {'scientificName': 'Branchiostoma lanceolatum', 'commonName': 'Common lancelet', 'taxonId': 7740}, {'scientificName': 'Haliotis rubra', 'taxonId': 36100}], 'goTerms': [{'goId': 'GO:0005509', 'aspect': 'GO Molecular Function'}, {'goId': 'GO:0005544', 'aspect': 'GO Molecular Function'}]}}]\n",
      "  >done\n",
      "trying to fetch UniRef90\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "{'Server': 'nginx/1.17.7', 'Vary': 'Accept-Encoding, Accept, Accept-Encoding, X-UniProt-Release, X-API-Deployment-Date, User-Agent', 'Cache-Control': 'no-cache', 'Content-Type': 'application/json', 'Access-Control-Allow-Credentials': 'true', 'Content-Encoding': 'gzip', 'Access-Control-Expose-Headers': 'Link, X-Total-Results, X-UniProt-Release, X-UniProt-Release-Date, X-API-Deployment-Date', 'X-API-Deployment-Date': '20-September-2022', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Date': 'Fri, 30 Sep 2022 10:57:29 GMT', 'X-UniProt-Release': '2022_03', 'Link': '<https://rest.uniprot.org/idmapping/uniref/results/49127fc9c8bdf08819ccd68d6f73071390e26d3c?cursor=t49h2mo5cklw&size=500>; rel=\"next\"', 'X-Total-Results': '6452', 'Transfer-Encoding': 'chunked', 'Access-Control-Allow-Origin': '*', 'Connection': 'keep-alive', 'Access-Control-Allow-Methods': 'GET, PUT, POST, DELETE, PATCH, OPTIONS', 'Access-Control-Allow-Headers': 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization', 'X-UniProt-Release-Date': '03-August-2022'}\n",
      "Fetched: 500 / 6452\n",
      "Fetched: 1000 / 6452\n",
      "Fetched: 1500 / 6452\n",
      "Fetched: 2000 / 6452\n",
      "Fetched: 2500 / 6452\n",
      "Fetched: 3000 / 6452\n",
      "Fetched: 3500 / 6452\n",
      "Fetched: 4000 / 6452\n",
      "Fetched: 4500 / 6452\n",
      "Fetched: 5000 / 6452\n",
      "Fetched: 5500 / 6452\n",
      "Fetched: 6000 / 6452\n",
      "Fetched: 6452 / 6452\n",
      "[{'from': 'P26256', 'to': {'id': 'UniRef90_P26256', 'name': 'Cluster: Annexin-B12', 'updated': '2022-08-03', 'entryType': 'UniRef90', 'commonTaxon': {'scientificName': 'Hydra vulgaris', 'taxonId': 6087}, 'memberCount': 1, 'organismCount': 1, 'representativeMember': {'memberIdType': 'UniProtKB ID', 'memberId': 'ANX12_HYDVU', 'organismName': 'Hydra vulgaris (Hydra) (Hydra attenuata)', 'organismTaxId': 6087, 'sequenceLength': 316, 'proteinName': 'Annexin-B12', 'accessions': ['P26256'], 'uniref50Id': 'UniRef50_P26256', 'uniref100Id': 'UniRef100_P26256', 'uniparcId': 'UPI0000125B90', 'seed': True, 'sequence': {'value': 'MVVQGTVKPHASFNSREDAETLRKAMKGIGTDEKSITHILATRSNAQRQQIKTDYTTLFGKHLEDELKSELSGNYEAAALALLRKPDEFLAEQLHAAMKGLGTDENALIDILCTQSNAQIHAIKAAFKLLYKEDLEKEIISETSGNFQRLLVSMLQGGRKEDEPVNAAHAAEDAAAIYQAGEGQIGTDESRFNAVLATRSYPQLHQIFHEYSKISNKTILQAIENEFSGDIKNGLLAIVKSVENRFAYFAERLHHAMKGLGTSDKTLIRILVSRSEIDLANIKETFQAMYGKSLYEFIADDCSGDYKDLLLQITGH', 'length': 316, 'molWeight': 35109, 'crc64': '10599869CBA853EC', 'md5': '2BB1D3F1A9250CA373652F1198DBCE09'}}, 'seedId': 'P26256', 'memberIdTypes': ['UniProtKB Reviewed (Swiss-Prot)'], 'members': ['P26256,0'], 'organisms': [{'scientificName': 'Hydra vulgaris', 'commonName': 'Hydra', 'taxonId': 6087}], 'goTerms': [{'goId': 'GO:0005509', 'aspect': 'GO Molecular Function'}, {'goId': 'GO:0005544', 'aspect': 'GO Molecular Function'}]}}]\n",
      "  >done\n",
      "trying to fetch UniRef100\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "Retrying in 3s\n",
      "{'Server': 'nginx/1.17.7', 'Vary': 'Accept-Encoding, Accept, Accept-Encoding, X-UniProt-Release, X-API-Deployment-Date, User-Agent', 'Cache-Control': 'no-cache', 'Content-Type': 'application/json', 'Access-Control-Allow-Credentials': 'true', 'Content-Encoding': 'gzip', 'Access-Control-Expose-Headers': 'Link, X-Total-Results, X-UniProt-Release, X-UniProt-Release-Date, X-API-Deployment-Date', 'X-API-Deployment-Date': '20-September-2022', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Date': 'Fri, 30 Sep 2022 10:58:47 GMT', 'X-UniProt-Release': '2022_03', 'Link': '<https://rest.uniprot.org/idmapping/uniref/results/938b2263fccb1e5478fcabc4a75e82e28f17624d?cursor=t49h2mo5cklw&size=500>; rel=\"next\"', 'X-Total-Results': '6452', 'Transfer-Encoding': 'chunked', 'Access-Control-Allow-Origin': '*', 'Connection': 'keep-alive', 'Access-Control-Allow-Methods': 'GET, PUT, POST, DELETE, PATCH, OPTIONS', 'Access-Control-Allow-Headers': 'DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Authorization', 'X-UniProt-Release-Date': '03-August-2022'}\n",
      "Fetched: 500 / 6452\n",
      "Fetched: 1000 / 6452\n",
      "Fetched: 1500 / 6452\n",
      "Fetched: 2000 / 6452\n",
      "Fetched: 2500 / 6452\n",
      "Fetched: 3000 / 6452\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched: 3500 / 6452\n",
      "Fetched: 4000 / 6452\n",
      "Fetched: 4500 / 6452\n",
      "Fetched: 5000 / 6452\n",
      "Fetched: 5500 / 6452\n",
      "Fetched: 6000 / 6452\n",
      "Fetched: 6452 / 6452\n",
      "[{'from': 'P26256', 'to': {'id': 'UniRef100_P26256', 'name': 'Cluster: Annexin-B12', 'updated': '2022-08-03', 'entryType': 'UniRef100', 'commonTaxon': {'scientificName': 'Hydra vulgaris', 'taxonId': 6087}, 'memberCount': 1, 'organismCount': 1, 'representativeMember': {'memberIdType': 'UniProtKB ID', 'memberId': 'ANX12_HYDVU', 'organismName': 'Hydra vulgaris (Hydra) (Hydra attenuata)', 'organismTaxId': 6087, 'sequenceLength': 316, 'proteinName': 'Annexin-B12', 'accessions': ['P26256'], 'uniref50Id': 'UniRef50_P26256', 'uniref90Id': 'UniRef90_P26256', 'uniparcId': 'UPI0000125B90', 'seed': True, 'sequence': {'value': 'MVVQGTVKPHASFNSREDAETLRKAMKGIGTDEKSITHILATRSNAQRQQIKTDYTTLFGKHLEDELKSELSGNYEAAALALLRKPDEFLAEQLHAAMKGLGTDENALIDILCTQSNAQIHAIKAAFKLLYKEDLEKEIISETSGNFQRLLVSMLQGGRKEDEPVNAAHAAEDAAAIYQAGEGQIGTDESRFNAVLATRSYPQLHQIFHEYSKISNKTILQAIENEFSGDIKNGLLAIVKSVENRFAYFAERLHHAMKGLGTSDKTLIRILVSRSEIDLANIKETFQAMYGKSLYEFIADDCSGDYKDLLLQITGH', 'length': 316, 'molWeight': 35109, 'crc64': '10599869CBA853EC', 'md5': '2BB1D3F1A9250CA373652F1198DBCE09'}}, 'seedId': 'P26256', 'memberIdTypes': ['UniProtKB Reviewed (Swiss-Prot)'], 'members': ['P26256,0'], 'organisms': [{'scientificName': 'Hydra vulgaris', 'commonName': 'Hydra', 'taxonId': 6087}]}}]\n",
      "  >done\n",
      "> mapping with dataset\n",
      "  >ok<  \n"
     ]
    }
   ],
   "source": [
    "DATASET = builder.sequence.add_cluster_info(DATASET)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Conservation\n",
    "\n",
    "Per-position conservation is important to have information on the importance of each amino acid in the protein. If a residue is conserved it means that it can have a critical role for the protein.  \n",
    "For now we will use only one formula to calculate the conservation: The **shannon entropy**.  \n",
    "Shanon Entropy(Shannon, 1948) is possibly the most sensitive tool to estimate the diversity of a system (http://imed.med.ucm.es/Tools/svs_help.html#ref).  \n",
    "For a multiple protein sequence alignment the Shannon entropy H for every position of the alignment is:  \n",
    "$$H=-\\sum_{i=1}^{M} P_i\\,log_2\\,P_i$$\n",
    "\n",
    " - Where $M$ is the number of different amino acids \n",
    " - $P_i$ is the 'propability' to have a specific amino acid on a column (number of occurences of this amino acid on $M$)\n",
    "\n",
    "From the github package by Felix Francis: https://github.com/ffrancis/Multiple-sequence-alignment-Shannon-s-entropy/blob/master/msa_shannon_entropy012915.py\n",
    "\n",
    "- H ranges from 0 (only one base/residue in present at that position) to 4.322 (all 20 residues are equally represented in that position).  \n",
    "- Typically, positions with H >2.0 are considerered variable, whereas those with H < 2 are consider conserved.  \n",
    "- Highly conserved positions are those with H <1.0 (Litwin and Jores, 1992). A minimum number of sequences is however required (~100) for H to describe the diversity of a protein family.  \n",
    "\n",
    "*Shannon, C. E. (1948) The mathematical theory of communication. The Bell system Technical Journal, 27, 379-423 & 623-656.*  \n",
    "*Litwin S., Jores R. (1992) Shannon Information as a Measure of Amino Acid Diversity. In: Perelson A.S., Weisbuch G. (eds) Theoretical and Experimental Insights into Immunology. NATO ASI Series (Series H: Cell Biology), vol 66. Springer, Berlin, Heidelberg*\n",
    "\n",
    "**We will use 2 Shannon entropy calculation** :\n",
    " - The regular one. All amino acids are considered are \"unique\". $Entropy_{M = 21} \\in [0,4.33 \\pm 0.01]$ estimated on 10000 repetitions of $N=10000$ random draw amoung 20 amino acids + gaps.\n",
    " - The H10 one. Amino acids are regroupes according their type (see http://thegrantlab.org/bio3d/reference/entropy.html), $Entropy_{M = 10}\\in [0,3.1 \\pm 0.01]$ estimated on 10000 repetitions of $N=10000$ random draw amoung 10 groups.\n",
    "   `Hydrophobic/Aliphatic [V,I,L,M], Aromatic [F,W,Y], Ser/Thr [S,T], Polar [N,Q], Positive [H,K,R], Negative [D,E], Tiny [A,G], Proline [P], Cysteine [C], and Gaps [-,X].`  \n",
    "    \n",
    "**NOTE** : The conservation is normalized between 0 and 1, 0 is not conserved, 1 is fully conserved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'pepr2ds.builder.Builder' from '/Home/ii/alexanderp/tubiana_etal_2022/pepr2ds/builder/Builder.py'>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n",
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n",
      "INFO: Pandarallel will run on 4 workers.\n",
      "INFO: Pandarallel will use Memory file system to transfer data between the main process and workers.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8e4e18fd23db40ce99e4778c6aa4c06c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import importlib\n",
    "importlib.reload(builderEngine)\n",
    "builder = builderEngine.Builder(SETUP, recalculate = recalculation_widget.value, notebook = True)\n",
    "\n",
    "\n",
    "DATASET = builder.sequence.add_conservation(DATASET, gapcutoff=0.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Saving dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## size optimisation\n",
    "Last thing to do before saving, is optimizing the dataset size. This will improve computational time (and disk storage of course 🙂 )\n",
    "One way to do it, it's to simplify the variables type. For example, change `str` to `category`\" will help decrease the space in memory a lot for variables like `domain` or `resdidue_name`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> datatypes optimisation\n",
      "Size BEFORE optimization 4825.02 MB\n",
      "Size AFTER optimization 2778.26 MB\n",
      "57.58% of the original size\n"
     ]
    }
   ],
   "source": [
    "DATASET = builder.optimize_size(DATASET)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Remove duplicated residues\n",
    "\n",
    "I don't know why, but sometimes I have duplicated residues (the whole cathpdb is duplicated) so just in case, we remove the duplicates based on `'atom_number','atom_name','residue_name','residue_number','cathpdb','chain_id`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATASET = DATASET.drop_duplicates(subset=['atom_number','atom_name','residue_name','residue_number','cathpdb','chain_id'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final saving\n",
    "\n",
    "**FINALY!** We have our final dataset !!! \n",
    "We save it into 2 versions: Full (with all PDB atoms) and light (only with `CA` and `CB` atoms).  \n",
    "➡️  For now we don't need other atoms for analysis, just CA and CB are enough and it can make the analysis faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "builder.save_checkpoint_dataset(DATASET,\"DATASET_peprmint_allatoms_d25\")\n",
    "builder.save_checkpoint_dataset(DATASET.query(\"atom_name in ['CA','CB']\"),\"DATASET_peprmint_d25\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ANNEXIN',\n",
       " 'BAR',\n",
       " 'C1',\n",
       " 'C2',\n",
       " 'C2DIS',\n",
       " 'ENTH',\n",
       " 'FYVE',\n",
       " 'GLA',\n",
       " 'PH',\n",
       " 'PLA',\n",
       " 'PLD',\n",
       " 'PX',\n",
       " 'SEC14',\n",
       " 'SH2',\n",
       " 'START']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(DATASET.domain.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1h10A00' '2x18A00' '1eazA00' '2hthB00' '2dx5A00' '1rrpB00' '1n3hA00'\n",
      " '3ml4A02' '2dyqA00' '3hk0A02' '4k81A02' '4k17B01' '4k17C01' '4k17A01'\n",
      " '4k17D01' '5efxA00' '3pvlA04' '1pmsA00' '2dtcA00' '3mpxA02' '2yf0A01'\n",
      " '2rgnB02' '3f0wA00' '1ddmA00' '1wj1A01' '4gzuA03' '1foeA02' '4gzuA02'\n",
      " '1v5uA00' '1wgqA00' '2d9yA00' '2yryA00' '2dkpA01' '2rovA00' '4tyzA00'\n",
      " '1v89A00' '2coaA01' '2d9zA01' '2ec1A00' '1y5oA00' '2dn6A00' '2mfqA00'\n",
      " '1fhoA00' '1plsA00' '2ys1A00' '1wi1A01' '3f5rA00' '4khbB01' '1tqzA01'\n",
      " '2d9wA01' '1z87A01' '1pfjA00' '2cofA00' '2rloA00' '1v88A00' '1v5pA01'\n",
      " '1v61A00' '2cocA01' '1evhA00' '1qc6A00' '1egxA00' '2dhkA01' '2m38A00'\n",
      " '1v5mA00' '2codA01' '1xr0B01' '2kupA01' '1mkeA00' '1x1fA00' '2fjlA00'\n",
      " '1wg7A01' '2yt0A01' '1xx0A00' '1x05A00' '1x1gA00' '2p8vA00' '1ddvA00'\n",
      " '1i7aA00' '2qklA00' '3m1iB00' '4l6eA00' '1k5dB00' '1xkeA00' '1ntyA02'\n",
      " '3aj4A00' '2dhiA00' '2d9vA01' '1ddwA00' '4f7uP00' '1vu2300' '1wvhA00'\n",
      " '2dkqA01' '2oqbA00' '1u5dA00' '1u5eA02' '3so6A00' '5c6rA00' '2da0A00'\n",
      " '1maiA00' '2vrwB02' '4ioyX02' '4khaA02' '4z2mB02' '1tj6A00' '3syxA00'\n",
      " '2jp2A00' '3fehA02' '4k2oA01' '3a8pD01' '2v76A00' '1p5tA00' '2dlwA00'\n",
      " '3voqA00' '1w1dA00' '4emoA00' '1e5wA03' '1h4rA03' '2i1jA03' '1gc6A03'\n",
      " '3fehA03' '2z0qA02' '4hhvA00' '2d9xA01' '3pp2A00' '2p0dA00' '4gxbA03'\n",
      " '2cy4A00' '2j59M01' '2dhjA00' '3dxcA00' '1wguA01' '2yszA01' '1zc3B00'\n",
      " '4bbkA00' '2lkoA00' '5l81A00' '2ys3A00' '2ej8A00' '1x86G02' '1xcgA02'\n",
      " '3odoA02' '1lw3A01' '5c16A01' '1faoA00' '4wsfA00' '1btnA00' '1wjmA00'\n",
      " '1droA00' '1dynA00' '3zvrA03' '2cayA00' '4chmA00' '2vszA02' '2gcjA02'\n",
      " '4ifsA02' '2elbA02' '4h8sA02' '3fssA02' '3gypA02' '3gyoA02' '1aqcB00'\n",
      " '1aqcA00' '1x11B00' '1x11A00' '4dbbA00' '3suzA01' '4d0nB02' '2lg1A02'\n",
      " '2y8fA00' '2crfA01' '2dx1A03' '3u7dA03' '2d4qA02' '4yonA02' '2y7bA00'\n",
      " '3n7cA00' '4nswA02' '1mixA02' '3g9wA02' '2al6A03' '3rcpA00' '2kcjA00'\n",
      " '3nsuA00' '4chjA00' '1ntvA00' '1m7eA00' '1aweA00' '2k2jA00' '1bakA00'\n",
      " '3au4A04' '1gg3A01' '2he7A03' '1irsA00' '3qwmA00' '1qqgA01' '1fgyA00'\n",
      " '1u27A00' '4kaxB00' '2aehA03' '4ekuA03' '3o96A01' '1p6sA00' '3tfmA02'\n",
      " '3tcaA02' '4gmvA02' '4y94A00' '2lulA00' '3d8dA00' '2dfkA02' '3ulbA01'\n",
      " '1kz7A02' '3qbvB02' '1upqA00' '4wj7A00' '3ml4A01' '3wyfB00' '1j0wA00'\n",
      " '4b6hA00' '2lydA00' '1q67A01' '5lonB01' '3cxbB00' '4iapA01' '4dx8A00'\n",
      " '3tfmA01']\n"
     ]
    }
   ],
   "source": [
    "#Generate alignment file list for C2DIS domain (S95, because everything is just too slow, too much structure)\n",
    "\n",
    "def selectUniquePerCluster(df, cathCluster, Uniref, withAlignment=True, pdbreference=None,\n",
    "                               removeStrand=False):\n",
    "        \"\"\"\n",
    "        Return a datasert with only 1 data per choosed clusters.\n",
    "        \"\"\"\n",
    "\n",
    "        if cathCluster not in [\"S35\", \"S60\", \"S95\", \"S100\"]:\n",
    "            raise ValueError('CathCluster given not in [\"S35\",\"S60\",\"S95\",\"S100\"]')\n",
    "\n",
    "        if Uniref not in [\"uniref50\", \"uniref90\", \"uniref100\"]:\n",
    "            raise ValueError('CathCluster given not in [\"uniref50\",\"uniref90\",\"uniref100\"]')\n",
    "\n",
    "        if withAlignment:\n",
    "            df = df[~df.alignment_position.isnull()]\n",
    "\n",
    "        cathdf = df.query(\"data_type == 'cathpdb'\")\n",
    "        seqdf = df.query(\"data_type == 'prosite'\")\n",
    "\n",
    "        def selectUniqueCath(group):\n",
    "            uniqueNames = group.cathpdb.unique()\n",
    "            if pdbreference:\n",
    "                if pdbreference in uniqueNames:\n",
    "                    select = pdbreference\n",
    "                else:\n",
    "                    select = uniqueNames[0]\n",
    "            else:\n",
    "                select = uniqueNames[0]\n",
    "\n",
    "            # return group.query(\"cathpdb == @select\")\n",
    "            return select\n",
    "\n",
    "        def selectUniqueUniref(group, exclusion):\n",
    "            uniqueNames = group.uniprot_acc.unique()\n",
    "            select = uniqueNames[0]\n",
    "            # return group.query(\"uniprot_acc == @select\")\n",
    "            if select not in exclusion:\n",
    "                return select\n",
    "        dfReprCathNames = cathdf.groupby([\"domain\", cathCluster]).apply(selectUniqueCath).to_numpy()\n",
    "        print(dfReprCathNames)\n",
    "        excludeUniref = df.query(\"cathpdb in @dfReprCathNames\").uniprot_acc.unique()  # Structures are prior to sequences.\n",
    "        dfReprUnirefNames = seqdf.groupby([\"domain\", Uniref]).apply(selectUniqueUniref,\n",
    "                                                                    exclusion=excludeUniref).to_numpy()\n",
    "        dfReprCath = cathdf.query(\"cathpdb in @dfReprCathNames\")\n",
    "        dfReprUniref = seqdf.query(\"uniprot_acc in @dfReprUnirefNames\")\n",
    "\n",
    "        return (pd.concat([dfReprCath, dfReprUniref]))\n",
    "\n",
    "    \n",
    "c2dis = selectUniquePerCluster(DATASET.query(\"domain== 'PH'\"), 'S95', 'uniref90', withAlignment=False, pdbreference='2da0A00')\n",
    "#pdblist = c2dis.cathpdb.dropna().unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['atom_number', 'atom_name', 'residue_name', 'chain_id',\n",
       "       'residue_number', 'x_coord', 'y_coord', 'z_coord', 'occupancy',\n",
       "       'b_factor', 'sec_struc', 'sec_struc_full', 'prot_block',\n",
       "       'sasa_rel_dssp', 'ASA_res_freesasa_florian', 'RSA_freesasa_florian',\n",
       "       'ASA_total_freesasa', 'ASA_mainchain_freesasa',\n",
       "       'ASA_sidechain_freesasa', 'RSA_sidechain_freesasa',\n",
       "       'RSA_total_freesasa_tien', 'RSA_sidechain_freesasa_tien',\n",
       "       'sec_struc_segment', 'pdb', 'domain', 'cathpdb', 'chain', 'uniprot_acc',\n",
       "       'data_type', 'Experimental Method', 'convhull_vertex',\n",
       "       'co_insertable_neighbors', 'density', 'is_co_insertable',\n",
       "       'is_hydrophobic_protrusion', 'neighboursID', 'neighboursList',\n",
       "       'protrusion', 'LDCI', 'S35', 'S60', 'S95', 'S100', 'S100Count',\n",
       "       'resolution', 'uniprot_id', 'origin', 'residue_index',\n",
       "       'alignment_position', 'prositeName', 'prositeID', 'ali_range',\n",
       "       'location', 'CR:prositeID', 'taxon', 'CR:prositeName', 'uniref50',\n",
       "       'uniref90', 'uniref100', 'shannon', 'shannonH10', 'type'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET.query(\"atom_name == 'CA' and domain =='PH'\").columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = DATASET.query(\"atom_name == 'CA' and domain =='PH' and data_type == 'cathpdb'\")[[\"ASA_res_freesasa_florian\",\"RSA_freesasa_florian\",\"ASA_total_freesasa\",\"ASA_mainchain_freesasa\",\"ASA_sidechain_freesasa\",\"RSA_sidechain_freesasa\",\"RSA_total_freesasa_tien\",\"RSA_sidechain_freesasa_tien\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6f562939bf91b46b655e3a917bf756908169d96bc255520a8d7ed621cd1fd7da"
  },
  "kernelspec": {
   "display_name": "Python [conda env:peprmint] *",
   "language": "python",
   "name": "conda-env-peprmint-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
